linkerd/app/core/src/errors.rs

```diff
@@ -229,32 +235,81 @@ impl<RspB: Default + hyper::body::HttpBody> respond::Respond<http::Response<RspB
-                Ok(http::Response::builder()
+                let rsp = set_http_status(builder, &*error)
                     .version(self.version)
-                    .status(status)
                     .header(http::header::CONTENT_LENGTH, "0")
                     .body(ResponseBody::default())
-                    .expect("error response must be valid"))
+                    .expect("error response must be valid");
+                let status = rsp.status();
+                debug!(%status, version = ?self.version, "Handling error with HTTP response");
+                Ok(rsp)
             }
         }
     }
 }
 
-fn http_status(error: &(dyn std::error::Error + 'static)) -> StatusCode {
+fn set_l5d_error_header(error: &(dyn std::error::Error + 'static)) -> http::response::Builder {
+    let mut builder = http::Response::builder();
+    if let Some(HttpError { message, .. }) = error.downcast_ref::<HttpError>() {
+        builder.header(L5D_HTTP_ERROR_MESSAGE, HeaderValue::from_static(message))
+    } else if error.is::<ResponseTimeout>() {
+        builder.header(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_static("request timed out"),
+        )
+    } else if error.is::<ConnectTimeout>() {
+        builder.header(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_static("failed to connect"),
+        )
+    } else if let Some(e) = error.downcast_ref::<FailFastError>() {
+        builder.header(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_str(&e.to_string()).unwrap_or_else(|error| {
+                warn!(%error, "Failed to encode fail-fast error message");
+                HeaderValue::from_static("service in fail-fast")
+            }),
+        )
+    } else if error.is::<tower::timeout::error::Elapsed>() {
+        builder.header(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_static("proxy dispatch timed out"),
+        )
+    } else if error.is::<IdentityRequired>() {
+        if let Ok(msg) = HeaderValue::from_str(&error.to_string()) {
+            builder = builder.header(L5D_HTTP_ERROR_MESSAGE, msg)
+        }
+        builder
+    } else if let Some(source) = error.source() {
+        set_l5d_error_header(source)
+    } else {
+        builder.header(
+            L5D_HTTP_ERROR_MESSAGE,
+            HeaderValue::from_static("proxy received invalid response"),
+        )
+    }
+}
+
+fn set_http_status(
+    builder: http::response::Builder,
+    error: &(dyn std::error::Error + 'static),
+) -> http::response::Builder {
     if let Some(HttpError { http, .. }) = error.downcast_ref::<HttpError>() {
-        *http
+        builder.status(*http)
     } else if error.is::<ResponseTimeout>() {
-        http::StatusCode::GATEWAY_TIMEOUT
-    } else if error.is::<FailFastError>() || error.is::<tower::timeout::error::Elapsed>() {
-        http::StatusCode::SERVICE_UNAVAILABLE
+        builder.status(http::StatusCode::GATEWAY_TIMEOUT)
```

Comment:
i.e. and then shorten this to just `StatusCode::...`

not a big deal either way, i guess